### DISCUSSION TITLE
What would you like to see in Philadelphia 2.0.0?

### DISCUSSION THREAD

**Discussion by jvirtanen** [paritytrading/philadelphia#170] [Ideas]
We released Philadelphia 1.0.0 back in 2018. Since then, we have kept improving the project mainly under the hood with a focus on performance.

The next major release, Philadelphia 2.0.0, will allow us to clean up and otherwise improve the library interface. What kind of changes would you like to see in Philadelphia 2.0.0?

**Comment by jheusser** [comment]
It could be useful if there was a way to provide default values when a `FIXValue` is `null`, instead of hand-crafting the `null` check, such as: 

```
long lastShares = message.valueOf(LastShares).asInt(0);
```

Obviously this code would not work right now the way `valueOf` works but just as example.

**Reply by jvirtanen** [reply]
This is an interesting idea. Perhaps a [sentinel value](https://en.wikipedia.org/wiki/Sentinel_value) of some sort could be returned instead of `null` for an unknown field. 🤔

**Reply by jvirtanen** [reply]
#219 proposes a solution to this.

**Comment by jvirtanen** [comment]
I would like to update the [`FIXMessageListener`](https://github.com/paritytrading/philadelphia/blob/1.2.0/libraries/core/src/main/java/com/paritytrading/philadelphia/FIXMessageListener.java) interface to the following form:
```java
/**
 * Receive a message.
 *
 * @param connection the connection
 * @param message a message
 * @throws IOException if an I/O error occurs
 */
void message(FIXConnection connection, FIXMessage message) throws IOException;
```

The addition of the `connection` parameter as well as enhancements in the public [`FIXConnection`](https://github.com/paritytrading/philadelphia/blob/1.2.0/libraries/core/src/main/java/com/paritytrading/philadelphia/FIXConnection.java) interface would make the [the built-in administrative message handling](https://github.com/paritytrading/philadelphia/blob/83e0af018f2fb599d9a2645979676ee69673f5b5/libraries/core/src/main/java/com/paritytrading/philadelphia/FIXConnection.java#L514) optional so that an application could provide its own, either simpler or more complete, implementation.

**Reply by barorion** [reply]
Supporting custom administrative message handling would be very useful, e.g. for applications that has a dedicated thread for incoming messages handling, and another dedicated thread for sending outgoing messages. The built in administrative message handling does not support such design.

**Comment by jheusser** [comment]
A way to provide a one-pass extraction of all `FIXValue` instances for a set of tags could also be nice. Currently there is a lot of unnecessary iteration over the same message to extract all the tags necessary via `FIXMessage#valueOf`.

**Reply by jvirtanen** [reply]
This is a good point, and I believe something like this can be already implemented on top of the current API.

[The Parity FIX Gateway source code](https://github.com/paritytrading/parity/blob/9bbfe78915da8e2f390e779e44fd51c1efbde1d5/applications/fix/src/main/java/com/paritytrading/parity/fix/Session.java#L132) contains an example of a single-pass `FIXMessage` parser. I think something along these lines could be extracted into a reusable class.

**Comment by spoerri** [comment]
How about rudimentary high availability? E.g. a standby node that could be manually failed over to

**Reply by jvirtanen** [reply]
Philadelphia does not currently implement any applications itself but rather only provides a library to help implementing applications. What would be the role of the node you are describing? 🙂

**Reply by jvirtanen** [reply]
By the way, [Artio](https://github.com/real-logic/artio) could be another open source project that you might be interested in. It is a FIX protocol implementation as a gateway application (that speaks [SBE](https://github.com/real-logic/simple-binary-encoding) to business applications) rather than as a library.

**Comment by tfmark** [comment]
Does Philadelphia support nested/repeating groups?

**Reply by jvirtanen** [reply]
It supports repeating groups in the sense that both inbound and outbound messages can contain them. To read or write them, however, the application has to do its own bookkeeping.

**Reply by mkst** [reply]
I imagine supporting repeating groups is a magnitude more complicated than supporting 'flat' messages -- and if there is no demand for it, then why complicate things. 

**EDIT** Realised I was logged in with my work account when posting the original message!

**Comment by mhartememx** [comment]
MessageHandler.handleResendRequest() in FIXConnection. does not correctly handle the case where the resend request has EndSeqNo == 0.

**Reply by jvirtanen** [reply]
Thanks! I opened a separate issue on this. 👍🏼

**Reply by jvirtanen** [reply]
I opened the pull request #185 to fix this. Just as a reminder, the plan is to have the option of custom handling for administrative messages besides the current built-in handling in Philadelphia 2.0.0. Nevertheless, I’m certainly interested to hear about and fix any protocol level quirks like this in the built-in handling. 🙂
